Force torque sensor broadcaster

destogl opened the pull request.

destogl commented on force_torque_sensor_controller/include/force_torque_sensor_controller/force_torque_sensor_controller.hpp:
```diff
@@ -0,0 +1,73 @@
+
+protected:
+  std::string sensor_name_;
+  std::vector<std::string> state_interface_names_;
```
```suggestion
  std::vector<std::string> interface_names_;
```

destogl commented on force_torque_sensor_controller/include/force_torque_sensor_controller/force_torque_sensor_controller.hpp:
```diff
@@ -0,0 +1,73 @@
+  std::string frame_id_;
+  int fx_range_;
+  int tz_range_;
+  std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::WrenchStamped>> sensor_state_publisher_;
```
Please use RealtimePublisher. (Check JointTrajectoryController)

bailaC replied:
Updated with RealtimePublisher. Kindly review and provide feedback.

destogl replied:
Can you then delete this one?

bailaC replied:
Yes. It is removed now.

destogl commented on force_torque_sensor_controller/src/force_torque_sensor_controller.cpp:
```diff
@@ -0,0 +1,171 @@
+    node->declare_parameter<std::string>("sensor_name", "");
+    node->declare_parameter<std::vector<std::string>>("state_interface_names", {});
+    node->declare_parameter<std::string>("frame_id", "");
+    node->declare_parameter<int>("fx_range", -1);
+    node->declare_parameter<int>("tz_range", -1);
```
```suggestion
```

destogl commented on force_torque_sensor_controller/src/force_torque_sensor_controller.cpp:
```diff
@@ -0,0 +1,171 @@
+  try {
+    auto node = get_node();
+    node->declare_parameter<std::string>("sensor_name", "");
+    node->declare_parameter<std::vector<std::string>>("state_interface_names", {});
```
```suggestion
    node->declare_parameter<std::vector<std::string>>("interface_names", {});
```

destogl commented on force_torque_sensor_controller/src/force_torque_sensor_controller.cpp:
```diff
@@ -0,0 +1,171 @@
+CallbackReturn ForceTorqueSensorController::on_configure(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  state_interface_names_ = node_->get_parameter("state_interface_names").as_string_array();
+  if (state_interface_names_.empty()) {
+    RCLCPP_ERROR(get_node()->get_logger(), "'state_interface_names' parameter was empty");
+    return CallbackReturn::ERROR;
+  }
```
```suggestion
  interface_names_ = node_->get_parameter("interface_names").as_string_array();
  if (interface_names_.empty()) {
    RCLCPP_ERROR(get_node()->get_logger(), "'interface_names' parameter was empty");
    return CallbackReturn::ERROR;
  }
```

destogl commented on force_torque_sensor_controller/src/force_torque_sensor_controller.cpp:
```diff
@@ -0,0 +1,171 @@
+    return CallbackReturn::ERROR;
+  }
+
+  sensor_name_ = node_->get_parameter("sensor_name").as_string();
```
put this first, to keep order: sensor_name, interface_names, frame_id

bailaC replied:
Done!

destogl commented on force_torque_sensor_controller/src/force_torque_sensor_controller.cpp:
```diff
@@ -0,0 +1,171 @@
+
+  try {
+    // register ft sensor data publisher
+    sensor_state_publisher_ = get_node()->create_publisher<geometry_msgs::msg::WrenchStamped>(
```
RealtimePublisher

destogl commented on force_torque_sensor_controller/src/force_torque_sensor_controller.cpp:
```diff
@@ -0,0 +1,171 @@
+  } catch (...) {
+    return CallbackReturn::ERROR;
+  }
+
```
Here you can also initialize the message.

bailaC replied:
Added the initialization.

destogl commented on force_torque_sensor_controller/src/force_torque_sensor_controller.cpp:
```diff
@@ -0,0 +1,171 @@
+  }
+
+  wrench_state_msg_.header.stamp = get_node()->get_clock()->now();
+  wrench_state_msg_.header.frame_id = frame_id_;
```
add this into configure.

bailaC replied:
I think, line number 153 can be added inside configure. But line 152 should be here as it will be different every time it publishes the message. Please suggest.

destogl commented on force_torque_sensor_controller/src/force_torque_sensor_controller.cpp:
```diff
@@ -0,0 +1,171 @@
+
+controller_interface::return_type ForceTorqueSensorController::update()
+{
+  geometry_msgs::msg::Vector3 f_vec;
```
Why not work directly with the wrench message?

bailaC replied:
I made some modifications. Please check once and suggest if this is ok.

bailaC commented on force_torque_sensor_controller/include/force_torque_sensor_controller/force_torque_sensor_controller.hpp:
```diff
@@ -0,0 +1,73 @@
+#include <string>
+#include <vector>
+
+#include "controller_interface/controller_interface.hpp"
```
Arranged like below,

```
#include <string>
#include <vector>

#include "controller_interface/controller_interface.hpp"
#include "force_torque_sensor_controller/visibility_control.h"
#include "geometry_msgs/msg/wrench_stamped.hpp"
#include "rclcpp_lifecycle/node_interfaces/lifecycle_node_interface.hpp"
#include "rclcpp_lifecycle/state.hpp"
```

bmagyar commented on force_torque_sensor_controller/include/force_torque_sensor_controller/force_torque_sensor_controller.hpp:
```diff
@@ -0,0 +1,68 @@
+
+#include "controller_interface/controller_interface.hpp"
+#include "force_torque_sensor_controller/visibility_control.h"
+#include "geometry_msgs/msg/wrench_stamped.hpp"
```
please make sure to declare this as a dependency in the cmake and in the package.xml as well

bailaC replied:
Dependency has been added. Please check once and provide feedback.

destogl commented on force_torque_sensor_controller/package.xml:
```diff
@@ -0,0 +1,31 @@
+  <depend>rclcpp</depend>
+  <depend>rclcpp_lifecycle</depend>
+  <depend>realtime_tools</depend>
+
```
```suggestion
```

destogl commented on force_torque_sensor_controller/include/force_torque_sensor_controller/force_torque_sensor_controller.hpp:
```diff
@@ -0,0 +1,76 @@
+  std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::WrenchStamped>> sensor_state_publisher_;
+  geometry_msgs::msg::WrenchStamped wrench_state_msg_;
+
+  using ControllerStateMsg = geometry_msgs::msg::WrenchStamped;
```
I would avoid this usings. There are only a few lines where explicit types are needed and these definitions do really increase readability.

bailaC replied:
Removed the using. Kind of following the old template! :)

destogl commented on force_torque_sensor_controller/include/force_torque_sensor_controller/force_torque_sensor_controller.hpp:
```diff
@@ -0,0 +1,76 @@
+  using ControllerStateMsg = geometry_msgs::msg::WrenchStamped;
+  using StatePublisher = realtime_tools::RealtimePublisher<ControllerStateMsg>;
+  using StatePublisherPtr = std::unique_ptr<StatePublisher>;
+  rclcpp::Publisher<ControllerStateMsg>::SharedPtr publisher_;
```
Can the normal publisher be deleted?

bailaC replied:
Done.

destogl commented on force_torque_sensor_controller/src/force_torque_sensor_controller.cpp:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  // publish
+  sensor_state_publisher_->publish(wrench_state_msg_);
```
Why not remove this publisher?

destogl replied:
Why are you publishing twice?

bailaC replied:
Corrected.

destogl reviewed (commented):
Looking good. Just a few more comments.

destogl commented on force_torque_sensor_controller/include/force_torque_sensor_controller/force_torque_sensor_controller.hpp:
```diff
@@ -0,0 +1,74 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
If you like, you can add

```suggestion

// Author: Subhas Das
```

Karsten1987 commented:
Can you elaborate a little bit on what's a FTSensorController is doing? What's the main purpose of this? What does it control?

Wouldn't it be more sensible to develop the capabilities of the FTSensor as part of the semantic components as described here? https://github.com/ros-controls/roadmap/blob/master/design_drafts/hardware_access.md#controller-interface

v-lopez commented:
I'm sorry I did not see this before.

I'm adding the state interfaces for IMU and FT sensor on Gazebo here: https://github.com/ros-simulation/gazebo_ros2_control/pull/65

In it, I'm using different interface names for the FT.
You are using `fx fy fz tx ty tz` and I'm using more verbose names `force.x force.y force.z torque.x torque.y torque.z`.
 
While for the FT sensor I have no preference for either, for the IMU finding such short names is harder `ox oy oz ow lax lay laz avx avy avz` would be quite confusing, hence why I went with the longer names there, and also for the FT state interface.

I actually brought this up in the last meeting but unfortunately @destogl could not attend.

Should we decide on a unique naming for this? I'm up for changing my PR, I just want a uniform naming scheme for all interfaces.

Karsten1987 commented:
I like more verbose names in this case. But yet again, I don't understand why this is implemented as a controller. With @v-lopez Gazebo use case, it makes even more sense to me to have these sensor packages implemented as individual software packages and being reused across any controller and Gazebo.

v-lopez commented:
@Karsten1987 "as individual software packages" implies regular packages unrelated to ros control, right? And the communication with the controllers is via ROS2? 

In our use case, the imu or ft are connected to the same ethercat or can bus as the actuators, and some of these devices run at 1 or 2khz. So ros control is a great way of reading them with minimal latency and in the same control cycle.

Karsten1987 commented:
Olivier kind of beat me to my answer. 

I might miss a high-level picture here, but from what I understood this F/T (or IMU for that matter) sensor implementation should serve as a convenience function to easily access the data in a more user-friendly API. Maybe even have some general functions on it to smooth data etc. These convenience function can then be used within a controller - but they aren't a controller.

What's coming in parallel to it is what Olivier just mentioned. There should be a FT sensor implementation, but on hardware level. That's only to say where the data comes from and fills up the resource interfaces, i.e. the 6D data for the FT sensor. That way you still have the data accessible in each update cycle and can achieve your ethercat bus update speed.

destogl commented:
@Karsten1987 I just added the first semantic component into ros-controls/ros2_control#370. Can you check and comment on the implementation?

@bailaC I corrected your tests, but now moving to semantic components, they are broken again. Can you look at this? And please investigate different combinations of parameters as described their use in doc/userdoc.rst.

Irritation: `ament_cpplint` is happy only my computer...

bmagyar commented on force_torque_sensor_broadcaster/CMakeLists.txt:
```diff
@@ -0,0 +1,120 @@
+  )
+endif()
+
+ament_target_dependencies(test_load_force_torque_sensor_broadcaster
```
umm... this statement managed to slip outside of the `if()`

bailaC replied:
Corrected.

bmagyar commented on force_torque_sensor_broadcaster/include/force_torque_sensor_broadcaster/force_torque_sensor_broadcaster.hpp:
```diff
@@ -0,0 +1,83 @@
+// limitations under the License.
+
+/*
+ * Author: Subhas Das, Denis Stogl
```
authors are displayed twice

bailaC replied:
Removed one of them.

bmagyar commented on force_torque_sensor_broadcaster/package.xml:
```diff
@@ -0,0 +1,30 @@
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>force_torque_sensor_broadcaster</name>
+  <version>0.0.0</version>
```
please set the version to the version of the latest release from this repo

bmagyar commented on force_torque_sensor_broadcaster/package.xml:
```diff
@@ -0,0 +1,30 @@
+  <name>force_torque_sensor_broadcaster</name>
+  <version>0.0.0</version>
+  <description>Controller to publish state of force-torque sensors.</description>
+  <maintainer email="denis@stogl.de">Denis Štogl</maintainer>
```
```suggestion
  <maintainer email="bence.magyar.robotics@gmail.com">Bence Magyar</maintainer>
  <maintainer email="denis@stogl.de">Denis Štogl</maintainer>
```

bmagyar commented on force_torque_sensor_broadcaster/package.xml:
```diff
@@ -0,0 +1,30 @@
+  <depend>realtime_tools</depend>
+
+  <test_depend>ament_lint_auto</test_depend>
+  <test_depend>ament_lint_common</test_depend>
```
is this required? it pulls in a bunch of linters we aren't using in this package

bailaC replied:
removed

bmagyar commented on force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:
```diff
@@ -0,0 +1,175 @@
+  interface_names_[3] = node_->get_parameter("interface_names.torque.x").as_string();
+  interface_names_[4] = node_->get_parameter("interface_names.torque.y").as_string();
+  interface_names_[5] = node_->get_parameter("interface_names.torque.z").as_string();
+
+  if (sensor_name_.empty() &&
+    std::count(interface_names_.begin(), interface_names_.end(), "") == 6)
+  {
+    RCLCPP_ERROR(
+      get_node()->get_logger(), "'sensor_name' or at least one "
+      "'interface_names.[force|torque].[x|y|z]' parameter has to be specified.");
+    return CallbackReturn::ERROR;
+  }
+
+  if (!sensor_name_.empty() &&
+    std::count(interface_names_.begin(), interface_names_.end(), "") != 6)
+  {
```
let's save re-counting those `""`-s and make the code a bit more readable
```suggestion

  const bool no_interface_names_defined = std::count(interface_names_.begin(), interface_names_.end(), "") == 6;

  if (sensor_name_.empty() && no_interface_names_defined)
  {
    RCLCPP_ERROR(
      get_node()->get_logger(), "'sensor_name' or at least one "
      "'interface_names.[force|torque].[x|y|z]' parameter has to be specified.");
    return CallbackReturn::ERROR;
  }

  if (!sensor_name_.empty() && !no_interface_names_defined)
  {
```

bmagyar commented on force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:
```diff
@@ -0,0 +1,175 @@
+    sensor_state_publisher_ = node_->create_publisher<geometry_msgs::msg::WrenchStamped>(
+      "~/sensor_states", rclcpp::SystemDefaultsQoS());
+    realtime_publisher_ = std::make_unique<StatePublisher>(sensor_state_publisher_);
+  } catch (...) {
```
is there a class of exceptions that we could capture here? it'd be nice to have an error print, otherwise poor old soul whose broadcaster fails here won't know what's happening

bailaC replied:
Added below exception in this place.

```
  } catch (const std::exception & e) {
    fprintf(stderr, "Exception thrown during publisher creation at configure stage with message
      : %s \n", e.what());
    return CallbackReturn::ERROR;
  }
```

bmagyar commented on force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:
```diff
@@ -0,0 +1,175 @@
+  realtime_publisher_->msg_.header.frame_id = frame_id_;
+  realtime_publisher_->unlock();
+
+  RCLCPP_INFO_STREAM(get_node()->get_logger(), "configure successful");
```
```suggestion
  RCLCPP_DEBUG_STREAM(get_node()->get_logger(), "configure successful");
```

bmagyar commented on force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:
```diff
@@ -0,0 +1,175 @@
+{
+  controller_interface::InterfaceConfiguration state_interfaces_config;
+  state_interfaces_config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+
```
```suggestion
```

bmagyar commented on force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:
```diff
@@ -0,0 +1,175 @@
+  state_interfaces_config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+
+  state_interfaces_config.names = force_torque_sensor_->get_state_interface_types();
+
```
```suggestion
```

bailaC replied:
done.

bmagyar commented on force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:
```diff
@@ -0,0 +1,175 @@
+{
+  controller_interface::InterfaceConfiguration command_interfaces_config;
+  command_interfaces_config.type = controller_interface::interface_configuration_type::NONE;
+
```
```suggestion
```

bmagyar commented on force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:
```diff
@@ -0,0 +1,175 @@
+CallbackReturn ForceTorqueSensorBroadcaster::on_activate(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  RCLCPP_INFO(node_->get_logger(), "Number of state interface is: %d.", state_interfaces_.size());
```
```suggestion
  RCLCPP_DEBUG(node_->get_logger(), "Number of state interface is: %d.", state_interfaces_.size());
```

bmagyar commented on force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:
```diff
@@ -0,0 +1,175 @@
+{
+  if (realtime_publisher_ && realtime_publisher_->trylock()) {
+    realtime_publisher_->msg_.header.stamp = node_->now();
+    RCLCPP_INFO(node_->get_logger(), "Trying to get values");
```
```suggestion
    RCLCPP_DEBUG(node_->get_logger(), "Trying to get values");
```

bmagyar commented on force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:
```diff
@@ -0,0 +1,175 @@
+    realtime_publisher_->msg_.header.stamp = node_->now();
+    RCLCPP_INFO(node_->get_logger(), "Trying to get values");
+    realtime_publisher_->msg_.wrench = force_torque_sensor_->get_values_as_message();
+    RCLCPP_INFO(node_->get_logger(), "Aftert getting values");
```
```suggestion
    RCLCPP_DEBUG(node_->get_logger(), "Aftert getting values");
```

bailaC replied:
Changed to,
`RCLCPP_DEBUG(node_->get_logger(), "After getting values");`

bmagyar commented on force_torque_sensor_broadcaster/test/test_force_torque_sensor_broadcaster.cpp:
```diff
@@ -0,0 +1,311 @@
+ * Authors: Subhas Das, Denis Stogl
+ */
+
+#include <stddef.h>
```
I'd opt for removing this one but even if we *really* need to keep it, `cstddef` would be the name

https://en.wikibooks.org/wiki/C_Programming/stddef.h

bailaC replied:
Removed.

bmagyar commented on force_torque_sensor_broadcaster/test/test_force_torque_sensor_broadcaster.cpp:
```diff
@@ -0,0 +1,311 @@
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
```
wanna include `chrono`?

bailaC replied:
@bmagyar , can you please explain the reason. Honestly, I was thinking of cutting the header files whose absence is not hampering the compilation/functionality.

bmagyar commented:
All in all, nice showcasing of the semantic components idea! I only had a few minor points, most are directly applicable through github

destogl commented:
@bmagyar I just finalized the tests, rebased to master, and from my side this PR is finalized. We just need to wait for tests for semantic components in ros-controls/ros2_control#370.

Please take a look and approve is you are happy.

bmagyar commented:
@destogl does this need another update to sync with ros2_control master or ready as-is?

destogl commented:
> @destogl does this need another update to sync with ros2_control master or ready as-is?

It is, I will retrigger pipelines and adjust if anything is needed.

bailaC commented:
@bmagyar , I hope all of your requested changes are complete. Can you please resolve the 'changes requested' section?

Can someone restart the checks? Or please let me know the process to trigger it.

bmagyar reviewed (approved):
@destogl may merge when ready

destogl commented on force_torque_sensor_broadcaster/test/test_load_force_torque_sensor_broadcaster.cpp:
```diff
@@ -0,0 +1,44 @@
+    cm.load_controller(
+      "test_force_torque_sensor_broadcaster",
+      "force_torque_sensor_broadcaster/ForceTorqueSensorBroadcaster"));
+}
```
```suggestion

rclcpp::shutdown();
}
```

bailaC commented:
@bmagyar , @jordan-palacios, @destogl  can anyone merge this now?

bmagyar reviewed (approved):
Thanks for stabilizing those tests!
